### PR TITLE
test(services): add comprehensive unit tests for beta services

### LINKED ISSUE #92: [Phase 5a] Unit Tests for Beta Services
# Phase 5a: Unit Tests for Beta Services

**Parent Epic**: #79
**Depends on**: Phase 2 (#82-88), Phase 3 (#89-90), Phase 4 (#91)

## Overview

Implement comprehensive unit tests for all new beta services to ensure 80%+ code coverage and proper functionality.

## Testing Strategy

Follow the project's testing standards (per `.augment/rules/testing.md`):

1. **Test actual behavior** - not just pass trivially
2. **Never skip failing tests** - fix the underlying issue
3. **Use pytest** as the testing framework
4. **Mock external dependencies** - HTTP calls, etc.
5. **Cover edge cases** - validation errors, empty responses, etc.

## Services to Test

| Service | Endpoints | Test File |
|---------|-----------|----------|
| IncidentsService | 6 | `tests/unit/test_incidents.py` |
| EngagementsService | 6 | `tests/unit/test_engagements.py` |
| BrandsService | 5 | `tests/unit/test_brands.py` |
| UomsService | 6 | `tests/unit/test_uoms.py` |
| QuestionAnswersService | 5 | `tests/unit/test_question_answers.py` |
| RecommendationsService | 2 | `tests/unit/test_recommendations.py` |
| SearchService | 2 | `tests/unit/test_search.py` |
| SubscribersService | 3 | `tests/unit/test_subscribers.py` |
| PreferencesService | 2 | `tests/unit/test_preferences.py` |
| Other beta services | ~15 | Various test files |

## Test Structure

```python
# tests/unit/test_incidents.py
import pytest
from unittest.mock import Mock, patch
from devrev.services.incidents import IncidentsService
from devrev.models.incidents import (
    Incident, IncidentCreate, IncidentUpdate,
    IncidentStage, IncidentSeverity
)


class TestIncidentsService:
    """Unit tests for IncidentsService."""
    
    @pytest.fixture
    def mock_http(self):
        """Create a mock HTTP client."""
        return Mock()
    
    @pytest.fixture
    def service(self, mock_http):
        """Create service with mock HTTP client."""
        return IncidentsService(mock_http)
    
    def test_create_incident_success(self, service, mock_http):
        """Test creating an incident with valid data."""
        mock_http.post.return_value.json.return_value = {
            "incident": {
                "id": "don:core:...:incident/123",
                "title": "Test Incident",
                "stage": "acknowledged",
            }
        }
        mock_http.post.return_value.status_code = 200
        
        result = service.create(
            IncidentCreate(
                title="Test Incident",
                severity=IncidentSeverity.SEV1,
            )
        )
        
        assert result.id == "don:core:...:incident/123"
        assert result.title == "Test Incident"
        mock_http.post.assert_called_once()
    
    def test_create_incident_validation_error(self, service):
        """Test that empty title raises validation error."""
        with pytest.raises(ValueError):
            IncidentCreate(title="")  # Empty title should fail
    
    def test_list_incidents_with_filters(self, service, mock_http):
        """Test listing incidents with stage and severity filters."""
        mock_http.post.return_value.json.return_value = {
            "incidents": [
                {"id": "inc/1", "title": "Incident 1"},
                {"id": "inc/2", "title": "Incident 2"},
            ],
            "next_cursor": "cursor123",
        }
        mock_http.post.return_value.status_code = 200
        
        result = service.list(
            stage=[IncidentStage.ACKNOWLEDGED],
            severity=[IncidentSeverity.SEV1],
            limit=10,
        )
        
        assert len(result.incidents) == 2
        assert result.next_cursor == "cursor123"
    
    def test_get_incident_not_found(self, service, mock_http):
        """Test getting non-existent incident raises NotFoundError."""
        from devrev.exceptions import NotFoundError
        mock_http.get.side_effect = NotFoundError("Incident not found")
        
        with pytest.raises(NotFoundError):
            service.get("invalid-id")
    
    def test_update_incident_stage(self, service, mock_http):
        """Test updating incident stage."""
        mock_http.post.return_value.json.return_value = {
            "incident": {
                "id": "inc/123",
                "title": "Test",
                "stage": "resolved",
            }
        }
        mock_http.post.return_value.status_code = 200
        
        result = service.update(
            IncidentUpdate(
                id="inc/123",
                stage=IncidentStage.RESOLVED,
            )
        )
        
        assert result.stage == IncidentStage.RESOLVED
    
    def test_delete_incident_success(self, service, mock_http):
        """Test deleting an incident."""
        mock_http.post.return_value.status_code = 204
        mock_http.post.return_value.content = b""
        
        service.delete("inc/123")  # Should not raise
        mock_http.post.assert_called_once()
    
    def test_group_incidents(self, service, mock_http):
        """Test grouping incidents by field."""
        mock_http.post.return_value.json.return_value = {
            "groups": [
                {"key": "sev1", "count": 5},
                {"key": "sev2", "count": 10},
            ]
        }
        mock_http.post.return_value.status_code = 200
        
        result = service.group(group_by="severity")
        
        assert len(result.groups) == 2
```

## Tasks

- [ ] Create test files for each new service
- [ ] Implement tests for all CRUD operations
- [ ] Test pagination handling
- [ ] Test error handling (404, 400, 500)
- [ ] Test model validation
- [ ] Test enum values
- [ ] Test async service variants
- [ ] Achieve 80%+ coverage on new code

## Coverage Requirements

```bash
# Run tests with coverage
pytest tests/unit/ --cov=src/devrev --cov-report=html --cov-report=term-missing

# Check coverage threshold
pytest --cov=src/devrev --cov-fail-under=80
```

## Acceptance Criteria

- [ ] All new services have comprehensive unit tests
- [ ] 80%+ line coverage on new code
- [ ] All tests pass (`pytest` exits 0)
- [ ] Tests cover success and error cases
- [ ] Tests cover edge cases (empty lists, null values)
- [ ] Async variants tested
- [ ] No skipped tests without justification

## Files to Create

- `tests/unit/test_incidents.py`
- `tests/unit/test_engagements.py`
- `tests/unit/test_brands.py`
- `tests/unit/test_uoms.py`
- `tests/unit/test_question_answers.py`
- `tests/unit/test_recommendations.py`
- `tests/unit/test_search.py`
- `tests/unit/test_subscribers.py`
- `tests/unit/test_preferences.py`
- `tests/unit/test_content_templates.py`
- `tests/unit/test_custom_link_types.py`
- `tests/unit/test_event_sources.py`

---

**Estimated effort**: 2-3 days

### LINKED ISSUE #93: [Phase 5b] Integration and Backwards Compatibility Tests
# Phase 5b: Integration and Backwards Compatibility Tests

**Parent Epic**: #79
**Depends on**: Phase 5a (#92)

## Overview

Implement integration tests for beta API endpoints and comprehensive backwards compatibility tests to ensure existing functionality is not broken.

### Critical Requirement

**All existing client code must continue to work without modification.**

This means:
1. Default client behavior unchanged
2. All public API endpoints work as before
3. No breaking changes to method signatures
4. No breaking changes to model structures
5. Import paths unchanged

### Backwards Compatibility Test Suite

```python
# tests/integration/test_backwards_compatibility.py
import pytest
from devrev import DevRevClient, AsyncDevRevClient, DevRevConfig
from devrev.exceptions import DevRevError, NotFoundError, ValidationError


class TestBackwardsCompatibility:
    """Ensure all existing functionality works unchanged."""
    
    def test_default_client_initialization(self):
        """Default client should use public API."""
        client = DevRevClient()
        # Should not raise, should default to public API
        assert client.api_version.value == "public"
    
    def test_existing_import_paths(self):
        """All existing imports should still work."""
        # These imports must not break
        from devrev import DevRevClient, AsyncDevRevClient
        from devrev import DevRevConfig, get_config, configure
        from devrev.exceptions import (
            DevRevError, AuthenticationError, NotFoundError,
            ValidationError, RateLimitError, ServerError
        )
        from devrev.models.accounts import Account, AccountCreate
        from devrev.models.works import Work, WorkCreate
    
    def test_accounts_service_unchanged(self, client):
        """Accounts service API should be unchanged."""
        # These method signatures must not change
        assert hasattr(client.accounts, 'list')
        assert hasattr(client.accounts, 'get')
        assert hasattr(client.accounts, 'create')
        assert hasattr(client.accounts, 'update')
        assert hasattr(client.accounts, 'delete')
    
    def test_works_service_unchanged(self, client):
        """Works service API should be unchanged."""
        assert hasattr(client.works, 'list')
        assert hasattr(client.works, 'get')
        assert hasattr(client.works, 'create')
        assert hasattr(client.works, 'update')
        assert hasattr(client.works, 'delete')
    
    def test_all_existing_services_accessible(self, client):
        """All existing services should be accessible."""
        services = [
            'accounts', 'articles', 'code_changes', 'conversations',
            'dev_users', 'groups', 'links', 'parts', 'rev_users',
            'slas', 'tags', 'timeline_entries', 'webhooks', 'works'
        ]
        for service_name in services:
            assert hasattr(client, service_name), f"Missing service: {service_name}"
    
    def test_config_environment_variables_unchanged(self, monkeypatch):
        """Existing environment variables should still work."""
        monkeypatch.setenv("DEVREV_API_TOKEN", "test-token")
        monkeypatch.setenv("DEVREV_BASE_URL", "https://api.devrev.ai")
        monkeypatch.setenv("DEVREV_TIMEOUT", "60")
        monkeypatch.setenv("DEVREV_LOG_LEVEL", "DEBUG")
        
        config = DevRevConfig()
        assert config.timeout == 60
        assert config.log_level == "DEBUG"
    
    def test_exception_hierarchy_unchanged(self):
        """Exception hierarchy should be unchanged."""
        from devrev.exceptions import (
            DevRevError, AuthenticationError, ForbiddenError,
            NotFoundError, ValidationError, ConflictError,
            RateLimitError, ServerError, ServiceUnavailableError
        )
        
        # All should inherit from DevRevError
        assert issubclass(AuthenticationError, DevRevError)
        assert issubclass(NotFoundError, DevRevError)
        assert issubclass(ValidationError, DevRevError)
```

## Integration Tests for Beta API

```python
# tests/integration/test_beta_api.py
import pytest
from devrev import DevRevClient, APIVersion


@pytest.mark.integration
class TestBetaAPIIntegration:
    """Integration tests for beta API endpoints."""
    
    @pytest.fixture
    def beta_client(self):
        """Create a beta API client."""
        return DevRevClient(api_version=APIVersion.BETA)
    
    def test_incidents_list(self, beta_client):
        """Test listing incidents via beta API."""
        result = beta_client.incidents.list(limit=5)
        assert hasattr(result, 'incidents')
    
    def test_search_hybrid(self, beta_client):
        """Test hybrid search via beta API."""
        result = beta_client.search.hybrid(
            query="test",
            limit=5,
        )
        assert hasattr(result, 'results')
    
    def test_beta_service_on_public_api_raises(self):
        """Accessing beta service on public API should raise."""
        from devrev.exceptions import BetaAPIRequiredError
        
        client = DevRevClient()  # Default public API
        
        with pytest.raises(BetaAPIRequiredError):
            client.incidents.list()
```

## Tasks

- [ ] Create backwards compatibility test suite
- [ ] Test all existing import paths
- [ ] Test all existing service methods
- [ ] Test configuration compatibility
- [ ] Test exception hierarchy
- [ ] Create beta API integration tests
- [ ] Test beta service access guards
- [ ] Document any intentional changes

## Acceptance Criteria

- [ ] All backwards compatibility tests pass
- [ ] No existing tests broken
- [ ] Beta API integration tests pass
- [ ] Clear error messages for beta-only features
- [ ] Documentation of any edge cases

## Files to Create

- `tests/integration/test_backwards_compatibility.py`
- `tests/integration/test_beta_api.py`
- `tests/integration/test_api_version_selection.py`

---

**Estimated effort**: 1-2 days

### PR DESCRIPTION
**Pull request by mgmonteleone**
# Phase 5: Unit Tests for Beta Services

## Summary

This PR implements comprehensive unit tests for all 10 beta services (Issue #92) and backwards compatibility tests (Issue #93).

### Unit Tests for Beta Services (Issue #92)

| Service | Test File | Tests |
|---------|-----------|-------|
| IncidentsService | \ | 8 tests |
| EngagementsService | \ | 8 tests |
| BrandsService | \ | 7 tests |
| UomsService | \ | 13 tests |
| QuestionAnswersService | \ | 12 tests |
| RecommendationsService | \ | 14 tests |
| SearchService | \ | 6 tests |
| PreferencesService | \ | 10 tests |
| NotificationsService | \ | 10 tests |
| TrackEventsService | \ | 10 tests |

**Total: 98 new unit tests for beta services**

### Backwards Compatibility Tests (Issue #93)

Added \ with:
- Import compatibility tests (main module, exceptions, models, HTTP config)
- Exception hierarchy tests (all exceptions inherit from DevRevError)
- Client initialization tests (default, explicit token, config object)
- Public services accessibility tests (all 14 public services)
- Beta services guard tests (raises BetaAPIRequiredError on public client)
- Configuration compatibility tests (env vars, direct init, global configure)
- API version selection tests (default PUBLIC, explicit BETA)

**Total: 22 backwards compatibility tests**

## Test Results

\\\

## Quality Checks

- ✅ ruff check passed
- ✅ ruff format passed
- ✅ All 366 tests pass

## Files Changed

- \ - Added fixtures for all beta services
- \ - Added async tests for QuestionAnswers
- 10 new test files for beta services
- 1 new backwards compatibility test file

Closes #92
Closes #93

### CODE REVIEW

File: tests/unit/services/test_track_events.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+                name="user_login",
+                properties={"source": "web"},
+                user_id="don:identity:user:123",
+            )
+        ]
+        result = service.publish(events)
+
+        assert isinstance(result, TrackEventsPublishResponse)
+        assert result.success is True
+        assert result.count == 2
```
[the remark this answers is not in the document]
Comment (mgmonteleone) [reply tests/unit/services/test_track_events.py]: Fixed: Updated both sync and async test_publish_single_event tests to override the mock response data with count=1 to accurately reflect that a single event is being published. The assertions now correctly match the test behavior.
